clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java

```diff
@@ -0,0 +1,235 @@
+                    time,
+                    true,
+                    apiVersions,
+                    fetcherThrottleTimeSensor,
+                    logContext);
+            this.networkClient = new ConsumerNetworkClient(
+                    logContext,
+                    netClient,
+                    metadata,
+                    time,
+                    retryBackoffMs,
+                    config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                    heartbeatIntervalMs);
+        } catch (Exception e) {
+            // now propagate the exception
+            throw new KafkaException("Failed to construct background processor", e);
+        }
+    }
+
+    // VisibleForTesting
+    DefaultBackgroundThreadRunnable(Time time,
+                                    ConsumerConfig config,
+                                    BlockingQueue<ApplicationEvent> applicationEventQueue,
+                                    BlockingQueue<BackgroundEvent> backgroundEventQueue,
+                                    SubscriptionState subscriptions,
+                                    ConsumerMetadata metadata,
+                                    ConsumerNetworkClient client) {
+        this.time = time;
+        this.config = config;
+        this.log = LoggerFactory.getLogger(getClass());
+        this.applicationEventQueue = applicationEventQueue;
+        this.backgroundEventQueue = backgroundEventQueue;
+        this.subscriptions = subscriptions;
+        this.metadata = metadata;
+        this.networkClient = client;
+        this.metrics = new Metrics();
+    }
+
+    private void setConfig() {
+        this.retryBackoffMs = this.config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        this.clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
+        this.heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+    }
+
+    @Override
+    public void run() {
+        try {
+            log.debug("DefaultBackgroundThreadRunnable started");
+            while (running) {
+                pollOnce();
+                this.wait(retryBackoffMs);
+            }
+        } catch (Exception e) {
+            // TODO: Define fine grain exceptions here
+        } finally {
+            log.debug("DefaultBackgroundThreadRunnable closed");
+        }
+    }
+
+    /**
+     * Process event from a single poll
+     */
+    void pollOnce() {
+        this.inflightEvent = maybePollEvent();
+        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
+            // clear inflight event upon successful consumption
+            this.inflightEvent = Optional.empty();
+        }
+        networkClient.pollNoWakeup();
+    }
+
+    public Optional<ApplicationEvent> maybePollEvent() {
+        if (this.inflightEvent.isPresent()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
```

Comment:
i thought about it, but I think maybe we don't want to block here, as each loop is already waiting for a `retryBackoffMs`. I'm thinking, if the queue returns nothing, then maybe we give it another try in the subsequent loop?

Reply:
+1 to @philipnee 's comment. The main reason is that I think eventually the background thread would not only try to just take events from the queue, since even though there's no events incoming, it would still try to repeatedly send fetch request and handle fetch responses -- well, for now maybe we cannot accumulate more than one fetched response per partition, but we may lift that with a memory limit in the future --- so the `runOnce` pattern would likely be:

1. try to get some event from the queue, if yes, then handle it and potentially send the corresponding request;
2. if the queue is empty, then try to send a fetch request.
3. poll the network client in order to receive any responses ready-to-read on the socket.

In that case, we would not want to wait on the queue at all if there's none, but directly move on to step 2) in each iteration.